packages/react/src/components/AccountSettings/SetupTOTP/SetupTOTP.tsx

```diff
@@ -0,0 +1,180 @@
+import React from 'react';
+import QRCode from 'qrcode';
+
+import {
+  AmplifyUser,
+  getLogger,
+  getTotpCodeURL,
+  setupTOTP,
+  translate,
+  verifyTOTPToken,
+} from '@aws-amplify/ui';
+
+import { useAuth } from '../../../internal';
+import { View, Flex } from '../../../primitives';
+import { FormValues } from '../types';
+import { ComponentClassName } from '../constants';
+import {
+  ConfirmationCode,
+  CopySecretKey,
+  Error,
+  SecretKeyQRCode,
+  SubmitButton,
+} from './defaults';
+import { ConfigureTOTPProps, TotpSecret, VerifyTotpStatus } from './types';
+import { QR_CODE_DIMENSIONS } from './constants';
+
+const logger = getLogger('Auth');
+
+function SetupTOTP({
+  totpIssuer = 'AWSCognito',
+  totpUsername,
+  onSuccess,
+  onError,
+}: ConfigureTOTPProps): JSX.Element | null {
+  const [formValues, setFormValues] = React.useState<FormValues>({ code: '' });
+  const [totpSecret, setTotpSecret] = React.useState<TotpSecret>(null);
+  const [verifyTotpStatus, setVerifyTotpStatus] =
+    React.useState<VerifyTotpStatus>({
+      isVerifying: false,
+      errorMessage: null,
+    });
+
+  const { user, isLoading } = useAuth();
+
+  const hasInit = React.useRef(false);
+
+  const generateQRCode = React.useCallback(
+    async (currentUser: AmplifyUser): Promise<void> => {
+      try {
+        const secretKey = await setupTOTP(currentUser);
+        const username = totpUsername || currentUser?.username;
+        const totpCode = getTotpCodeURL(totpIssuer, username, secretKey);
+        const qrCode = await QRCode.toDataURL(totpCode);
+
+        setTotpSecret({ secretKey, qrCode });
+      } catch (e) {
+        logger.error(e);
+      }
+    },
+    [totpIssuer, totpUsername]
+  );
+
+  React.useEffect(() => {
+    if (user && !hasInit.current) {
```

Comment:
`user` has to be there because `user` is initially `null`. I think I introduced `hasInit` because of some remounting issue with parent component.

Let me try using `!totpSecret` check again. If it turns out to be problematic later when we use this component in `ConfigureMFA`, we can revert it.

Reply:
Done in f658ad2e6f4c28a4b1d48b3eb2ba5ece098b78db